File: VERT_FSFB3.m
function [c2, Psi, G, N2, Pmid] = VERT_FSFB3(N2,Pmid)
%function [c2, Psi, G, N2, Pmid] = VERT_FSFB2(N2,Pmid)
%
%    VERT_FSFB.m
%
%     Gabriel A. Vecchi - May 12, 1998
%     modified by Youran 
%     1. The modes are sorted from low to high
%     2. barotropic mode has been removed

%%%%%%%%%%%%%%%%
% 
%    Solves the discretized wave projection problem
%    given the vertical profiles of Temperature, Salinity, Pressure 
%    and the depth inteval length.
%
%  Uses the seawater function sw_bfrq to calculate N2.
%%%%%%%%%%%%%%%%
%
%    Arguments:
%       T = temperature vector at same depths as salinity and pressure.
%       S = salinity vector at same depths as temperature and pressure.
%       P = pressure vector at same depths as temperature and salinity.
%       Dz = length of depth interval in meters.
%%%%%%%%%%%%%%%%
%
%    Returns:
%	c2 = vector of square of the wavespeed.
%       Psi = matrix of eigenvectors (horizontal velocity structure functions).
%       G  =  matrix of integral of eigenvectors (vertical velocity structure functions).
%       N2 = Brunt-Vaisla frequency calculated at the midpoint pressures.
%       Pmid = midpoint pressures.
%%%%%%%%%%%%%%%%

%  Find N2 - get a M-1 sized vector, at the equator.
%[N2,crap,Pmid] = sw_bfrq(S,T,P,0);

% remove nan values
N2 = N2(~isnan(N2));% N2 = N2(N2~=0);
Pmid = Pmid(~isnan(Pmid));% Pmid = Pmid(Pmid~=0);

% check the lengths
if length(N2)~=length(Pmid)
    error('N2 and Pmid does not have same length');
end


for i = 1:length(N2)
   if N2(i) < 0
      N2(i) = min(abs(N2));
   end;
end;

% bdc: needs equally-spaced depths!
Dz= median(diff(Pmid));

% add a point for the surface
M = length(N2)+1;

%  Fill in D - the differential operator matrix.
%  Surface (repeat N2 from midpoint depth)
D(1,1) = -2/N2(1);
D(1,2) = 2/N2(1);
%  Interior
for i = 2 : M-1,
D(i,i-1) = 1/N2(i-1);
D(i,i) = -1/N2(i-1)-1/N2(i);
D(i,i+1) = 1/N2(i);
end
%  Bottom
D(M,M-1) = 2/N2(M-1);
D(M,M) = -2/N2(M-1);
D=-D./(Dz*Dz);
%bdc: no need for A?
% A = eye(M);

% Calculate generalized eigenvalue problem
% bdc: eigs gets top M-1
%[Psi,lambda] = eigs(D,[],M-1);
% use eig:
[Psi,lambda] = eig(D);

% extract eigenvalues
% lambda = diag(lambda); % diag returns a column. better if making lambda into a row so that
% it is consistent with eigenfunctions.
lambda = sum(lambda);


% remove BT mode (lambda = 1/(radius of deformation * f)^2   (m/s)^-2
% BT mode, rd = infinity. hence lambda = 0.
ind=find(lambda>=1.e-10);
lambda=lambda(ind);
Psi=Psi(:,ind);


% sort
[lambda,ind]=sort(lambda);
Psi=Psi(:,ind);

% Calculate square of the wavespeed.
c2=1./lambda;


% normalize Psi into the range between -1 and 1.
% Psi at surface and bottom level should be 1 or -1.
% Psi = fliplr(Psi);
% c2 = fliplr(c2);
for i=1:size(Psi,2)
  Psi(:,i) = Psi(:,i)/Psi(1,i); % or Psi(:,i) = Psi(:,i)/Psi(end,i);
end


% normalize?
G = INTEGRATOR(M,Dz)*Psi;

function [INT] = INTEGRATOR(M,Del)
%function [INT] = INTEGRATOR(M,Del)
%
%    INTEGRATOR.m
%
%     Gabriel A. Vecchi - June 7, 1998
%%%%%%%%%%%%%%%%
%    Generates and integration matrix.
%    Integrates from first point to each point.
%%%%%%%%%%%%%%%%

INT = tril(ones(M));
INT = INT - 0.5*(eye(M));
INT(:,1) = INT(:,1) - 0.5;
INT = INT*Del;

File: VERT_FSFB3.py
def  VERT_FSFB3(N2_uniform, Pmid_uniform):
    
    import numpy as np
    from numpy import linalg as LA
    import scipy
    import gsw
    
    '''
    VERT_FSFB3.py - Yu Gao, April 18, 2022 
    This function solves the discretized wave projection problem, given the vertical profiles of Temperature, Salinity, Pressure and the depth inteval length. Note that it wants REGULARLY spaced depths. The seawater function sw_bfrq can be used to calculate N2.
    
    # gsw-python package: https://github.com/TEOS-10/GSW-Python.
    # gsw can be installed from a clone of the repo using
    # $ pip install gsw
    
    # Arguments:
    # T = temperature vector at same depths as salinity and pressure.
    # S = salinity vector at same depths as temperature and pressure.
    # P = pressure vector at same depths as temperature and salinity.
    # Dz = length of depth interval in meters.
    # Returns:
    # c2 = vector of square of the wavespeed.
    # Psi = matrix of eigenvectors (horizontal velocity and pressure structure functions).
    # G  =  matrix of integral of eigenvectors (vertical velocity structure functions).
    # N2 = Brunt-Vaisla frequency squared calculated at the midpoint pressures.
    # Pmid = midpoint pressures.
    '''

    
    import numpy as np
    import gsw
    from numpy import linalg as LA
    
    for i in range(len(N2_uniform)):
        if N2_uniform[i] < 0:
            N2_uniform[i] = np.min(np.absolute(N2_uniform))
   
    Dz = np.median(np.diff(Pmid_uniform))
    Dzrng = np.max(np.diff(Pmid_uniform)) - np.min(np.diff(Pmid_uniform))
    
    if(Dzrng > 1e-3 * Dz):
        print('Dz is not constant!!')
    # else: print('Dz is constant.')
    
    # add a point for the surface
    # note that this means that Psi has one more depth than N2!!
    M = len(N2_uniform) + 1
    #  Fill in D - the differential operator matrix.
    #  Surface (repeat N2 from midpoint depth)
    D = np.zeros([M, M]) # Make a M x M zero array
    
    D[0, 0] = -2/N2_uniform[0]
    D[0, 1] = 2/N2_uniform[0]

    # Interior
    for i in range(1, M-1):
        D[i, i-1] = 1/N2_uniform[i-1]
        D[i, i] = -1/N2_uniform[i-1] - 1/N2_uniform[i]
        D[i, i+1] = 1/N2_uniform[i]

    # Bottom
    #D[M-1, M-2] = 1/N2_uniform[M-2]
    D[M-1, M-2] = 2/N2_uniform[M-2]
    D[M-1, M-1]  = -2/N2_uniform[M-2]
    D = -1 * D / (Dz * Dz)
    
    # BDC: D should be a tridiagonal matrix
    # Calculate generalized eigenvalue problem
    Lambda, Psi = LA.eig(D)
    # LA.eig() Compute the eigenvalues and right eigenvectors of a square array.
    # Lambda:    The eigenvalues, each repeated according to its multiplicity.
    # Psi: The normalized (unit "length") eigenvectors, such that the 
    # column ``Psi[:,i]`` is the eigenvector corresponding to the eigenvalue Lambda[i].
    
    # Remove Barotropic Mode 
    # Lambda = 1/(radius of deformation * f)^2 (m/s)^{-2}
    # BT mode, rd = infinity. hence lambda = 0.
    ind = np.where(Lambda >= 1e-11)
    Lambda = Lambda[ind]
    Psi = Psi[:, ind].reshape([D.shape[0], len(ind[0])])
    
    # Sort eigenvalues and eigenvectors
    idx = np.argsort(Lambda)
    Lambda_sorted = Lambda[idx]
    Psi_sorted = Psi[:,idx]
    c2 = 1 / Lambda_sorted # Modal Speed
    
    # Reference eigencevtor matrix to top value
    for i in range(Psi_sorted.shape[1]):
        Psi_sorted[:, i] = Psi_sorted[:, i] / Psi_sorted[0, i]
    
    # Generates an integration matrix
    INT = INTEGRATOR(M, Dz)
    
    # 
    G = np.matmul(INT, Psi_sorted)

    return c2, Psi, G, N2_uniform, Pmid_uniform


def INTEGRATOR(M, Dz): 
    
    """
    Translated from Gabriel A. Vecchi - June 7, 1998 by Yu Gao, - April 18, 2022 
    This function Generates an integration matrix, 
    and Integrates from first point to each point.
    """
    
    import numpy as np
    
    INT = np.tril(np.ones(M))
    INT = INT - 0.5*(np.eye(M))
    INT[:,0] = INT[:,0] - 0.5;
    INT = INT * Dz
    
    return INT